### ISSUE TITLE
Problem requiring modules and with Chai as Promised under karma test runner.

### ISSUE THREAD

**Issue by Astridax** [NativeScript/NativeScript#1956]
Version of Nativescript: 1.7.1

Hey there, I've hit a few issues when testing applications using the Karma that I wondered if I could have a second more experienced pair of eyes look at. 

The first is a less serious issue at the moment since I have managed to fix it by modifying the library involved. The issue is with the library Chai as Promised.

They have some code as shown [here](https://github.com/domenic/chai-as-promised/issues/148): designed to set certain variables dependent on the environment it is running under. The code falls through to the else block when running under Karma. 

Usually the module supports running under Karma along with the [karma-chai-as-promised](https://github.com/vlkosinov/karma-chai-as-promised) plugin. Since Karma usually runs under the browser, the browser has a self variable, whereas when running under Nativescript it does not. I was wondering if there were any global variables I could attach data to when running under the Nativescript test runner?

The second issue is a more overarching one. When running tests, I don't seem to be able to import NPM modules, unless I find an accompanying karma plugin. For example, doing:
`var sinon = require('sinon')` fails entirely unless I have the karma-sinon plugin. 

For example:
`Error: com.tns.NativeScriptException: Failed to find module: "sinon", relative to: /app/tns_modules/`

What if I want to stub a module that doesn't have a karma plugin. Requiring the module doesn't work.

These problems all happen when running `tns test android --emulator --watch`

Many thanks in advance.

**Comment by Astridax** [comment]
Okay so I managed to fix the problem with the chai-as-promised plugin as well as the similar problem found in [sinon-stub-promise](https://github.com/substantial/sinon-stub-promise).

The basic problem seems to be that these plugins expect that if they are not loaded from node (hence don't have the module var or module.exports), they are running in a browser and most browsers reveal self and window variables. In certain instances where they are not available (such as this) the plugins break. Others like [moment.js](https://github.com/moment/moment/blob/develop/moment.js) pass through a calling context such as a 'this' variable and attach the modules to that. This is what was required in this instance. I'll submit pull requests for the offending modules.

What I'd like to know is do the nativescript devs have any plans to implement some global variables when running tests under the companion app. A window and self variable would be helpful in this case, even if all they do is reference the this context.

The importing modules problem still exists though.

**Comment by enchev** [comment]
Hey @Astridax 

You can access the global context using `global`.

**Comment by hdeshev** [comment]
The way I usually fix this it to wrap the offending library in a module that sets up any globals the library needs, imports the library, and then cleans up. Here's how we do it in the nativescript-angular/tests project:

https://github.com/NativeScript/nativescript-angular/blob/master/tests/app/tests/test-config.ts

You could see that `test-config` module preserving the global chai `assert` as set up by the test runner (Angular2 overwrites it with its own.), and it also reexports the global as a module var, so you could later do `import {assert} from "./test-config";` in your code.

I don't think we should add a global `window` variable as that would break third party libs that use the `window` global as a suboptimal way to detect whether they run in the browser or not. Modules that need such globals to work can easily be wrapped using the technique I described above.

**Comment by Astridax** [comment]
Thanks for your response, sorry for my belated reply. I agree wholly that Nativescript shouldn't be modified to add a window global. In the end I forked the library and repaired it (before your answer). I sadly couldn't convince the owner to accept the pull request. So hopefully this will serve as an answer to anyone who comes across this in the future. 

I've noticed that your project also has a fairly complex grunt build file that seems to copy files in to bin/dist/modules. Is that how you've got around importing test npm modules without including them in the karma conf or having karma plugins for them?

**Comment by hdeshev** [comment]
> I've noticed that your project also has a fairly complex grunt build file that seems to copy files in to bin/dist/modules. Is that how you've got around importing test npm modules without including them in the karma conf or having karma plugins for them?

No, the NativeScript grunt file predates the karma test runner, and I believe it copies files around to make filtering files and packing npm modules easier.

**Comment by Astridax** [comment]
Well I have had problems importing certain libraries that were test only and getting them to work with the Karma runner. For Chai as Promised and Sinon I managed to get karma- plugins and set them as frameworks, however I failed to get proxyquire to work at all. Any thoughts on this would be great!

For now I've changed the way I'm testing, by writing my unit tests to run via mocha on my dev machine rather than the device, by way of proxyquire to require the nativescript modules that don't exist apart from the device. However this isn't mentioned much anywhere in the docs. This proves to be much faster and seems to give me the flexibility to use any npm testing module. For integration / e2e tests mocking is less of a concern, so can be run on the device.
